sim/battle.ts

```diff
@@ -1127,42 +1127,49 @@ export class Battle extends Dex.ModdedDex {
 			let maxTeamSize = teamLengthData && teamLengthData.battle;
 			this.add('teampreview' + (maxTeamSize ? '|' + maxTeamSize : ''));
 			if (!maxTeamSize) maxTeamSize = 6;
-			this.p1.maxTeamSize = maxTeamSize;
-			this.p2.maxTeamSize = maxTeamSize;
-			this.p1.currentRequest = 'teampreview';
-			p1request = {teamPreview: true, maxTeamSize, side: this.p1.getRequestData()};
-			this.p2.currentRequest = 'teampreview';
-			p2request = {teamPreview: true, maxTeamSize, side: this.p2.getRequestData()};
+			for (let i = 0; i < this.sides.length; i++) {
+				let side = this.sides[i];
+				side.maxTeamSize = maxTeamSize;
+				side.currentRequest = 'teampreview';
+				requests[i] = {teamPreview: true, maxTeamSize, side: side.getRequestData()};
+			}
 			break;
 
 		default: {
-			this.p1.currentRequest = 'move';
-			let activeData = this.p1.active.map(pokemon => pokemon && pokemon.getRequestData());
-			p1request = {active: activeData, side: this.p1.getRequestData()};
-
-			this.p2.currentRequest = 'move';
-			activeData = this.p2.active.map(pokemon => pokemon && pokemon.getRequestData());
-			p2request = {active: activeData, side: this.p2.getRequestData()};
+			for (let i = 0; i < this.sides.length; i++) {
+				let side = this.sides[i];
+				side.currentRequest = 'move';
+				let activeData = side.active.map(pokemon => pokemon && pokemon.getRequestData());
+				requests[i] = {active: activeData, side: side.getRequestData()};
+			}
 			break;
 		}
 		}
 
-		if (p1request) {
-			if (!this.supportCancel || !p2request) p1request.noCancel = true;
-			this.p1.emitRequest(p1request);
-		} else {
-			this.p1.emitRequest({wait: true, side: this.p1.getRequestData()});
+		let allRequestsMade = true;
+		for (const request of requests) {
+			if (!request) allRequestsMade = false;
 		}
-
-		if (p2request) {
-			if (!this.supportCancel || !p1request) p2request.noCancel = true;
-			this.p2.emitRequest(p2request);
-		} else {
-			this.p2.emitRequest({wait: true, side: this.p2.getRequestData()});
+		for (let i = 0; i < this.sides.length; i++) {
+			let side = this.sides[i];
+			let request = requests[i];
+			if (request) {
+				if (!this.supportCancel || allRequestsMade) request.noCancel = true;
+				side.emitRequest(request);
+			} else {
+				side.emitRequest({wait: true, side: side.getRequestData()});
+			}
 		}
 
-		if (this.p1.isChoiceDone() && this.p2.isChoiceDone()) {
-			throw new Error(`Choices are done immediately after a request`);
+		let choiceDone = false;
+		for (const side of this.sides) {
+			if (side.isChoiceDone()) {
+				if (choiceDone) {
```

Comment:
This checks for "at least two choices done" rather than the intended check "all choices done".

You should be looking for

```js
if (this.sides.every(side => side.isChoiceDone())) {
```